runtime/compiler/env/j9method.cpp

```diff
@@ -6605,6 +6605,14 @@ TR_ResolvedJ9Method::getResolvedDynamicMethod(TR::Compilation * comp, I_32 callS
       {
       TR_OpaqueMethodBlock * targetJ9MethodBlock = NULL;
       uintptr_t * invokeCacheArray = (uintptr_t *) callSiteTableEntryAddress(callSiteIndex);
+      // invokedynamic resolution can either result in a valid entry in the corresponding CallSite table slot if successful,
+      // or an exception object otherwise. The CallSite table entry is a two-element array containing the MemberName
+      // and appendix objects necessary for constructing a resolved invokedynamic adapter method call.
+      // The exception object would not be an array type object, and therefore it is sufficient to check if
+      // the entry obtained from the CallSite table is an array instance to determine if the resolution was successful.
+      if (!fej9()->isArrayObject(*invokeCacheArray))
```

Comment:
I am having second thoughts about this one. if `*invokeCacheArray` is truly a pointer to an object, what prevents GC from moving that object while we try to access it? We may need to hold VMAccess while executing `if (!fej9()->isArrayObject(*invokeCacheArray))` (extend the `VMAccessCriticalSection` which starts below).
There is also a comment below that "// this will not work in AOT or JITServer", so probably there is some guarantee that we cannot enter this path for JITServer.

Reply:
> if *invokeCacheArray is truly a pointer to an object, what prevents GC from moving that object while we try to access it?

I did not feel it was necessary to ensure VM access for this. My reasoning was, immediately after checking if the corresponding entry in the call site table had a valid object pointer using

https://github.com/eclipse-openj9/openj9/blob/ab5dfcfca290227c639a8d6877ca11773e723a42/runtime/compiler/env/j9method.cpp#L5915-L5919

We want to check whether that non-null object pointer cached in the call site table slot was an array or not. If that object pointer at `ramClass->callSites[callSiteIndex]` (i.e, `*invokeCacheArray`) is not pointing to a valid array object, then we should abort anyway. The slot is populated by an object that the VM's invokedynamic resolution mechanism allocates and (I'm speculating) the GC probably can not move. It's not a regular Java object. Once a call site table slot is populated (only if it was NULL to begin with), it is not modified again. What we do need VM access for however is accessing the individual elements in the `invokeCacheArray`, which is why we have the VM access being acquired right after this.

Based on my reasoning, do you still think it is necessary to put the isArrayObject check within the block with VM access acquired?

> There is also a comment below that "// this will not work in AOT or JITServer", so probably there is some guarantee that we cannot enter this path for JITServer.

This was true when this code was first written during the initial OpenJDK MethodHandle implementation, but since then JITServer became capable of supporting OpenJDK MHs and I can confirm that this code path is executed in JITServer now based on my testing. I will update the comment here to reflect the current state.

Reply:
Responding to your overall review comment:

> Moreover, the JITServer should never receive any object references. Thus, the new frontend method should just fatal assert if executed at the server rather than sending a message at the client.

Wouldn't this mean that JITServer loses support for any method containing `invokedynamic` bytecodes?

Reply:
If that array stored in one of the callsites is really special and cannot be moved by GC (maybe because it does not reside in the heap), then you don't need VM access for checking whether it's an array. Interesting though that the two elements of the array are protected by VM access.

Reply:
`fej9()->targetMethodFromMemberName()` has a FATAL_ASSERT in the JITServer implementation.
```
TR_OpaqueMethodBlock*
TR_J9ServerVM::targetMethodFromMemberName(uintptr_t memberName)
   {
   TR_ASSERT_FATAL(false, "targetMethodFromMemberName must not be called on JITServer");
   return NULL;
   }
   ```
   so if we ever entered that path while compiling at the server, the assert would be triggered.

Reply:
Another point: `getResolvedDynamicMethod()` has a different implementation for JITServer where the server asks the client to execute the corresponding method and then, with the answers received from the client, the server builds a resolvedMethod. Thus, the server does not need to access Java objects.

Reply:
Thank you for pointing those out, I had the wrong understanding of how things were being handled in JITServer mode. I will make the changes as you suggested.

I will also enforce VM access requirement for checking if object is array, so as not to cause confusion in the future for others, as there is no real benefit to doing this without VM access but lots of potential danger if this API was later used to check if heap objects were array types.